### PR TITLE
chore: convert to typescript

### PR DESCRIPTION
**Pull request by keithlayne**
## Description

Caveats:

1. I think this is probably not worth it. Probably jsdoc comments would be sufficient to improve the DX when working on this plugin.  That would be easy if you wanna close this.
2. I wouldn't normally lay out a TS project like this, but this is a tiny project and it lets you see diffs in the PR.

Notes:

1. The types that are in the repo are IMO wrong. The way to specify them would be `export = PgSimplifyInflectorPlugin`, and force users to either use the CJS style import if using TS, or turn on `esModuleInterop`.  Actually changing that would be a breaking change for anyone that didn't have `esModuleInterop` turned on.  (Not that big an issue, dealt with in TS but seems hacky).
2. The change of nullability in top-level collections in postgraphile 4.9.0 had an impact on the test snapshots from my previous PR - using 4.6.0 would generate a diff at this point. I bumped it in this branch for that reason.

## Performance impact

A couple of checks where TS wasn't convinced that the index field on regex matches were defined.

## Security impact

unkown

## Checklist

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [x] I've added tests for the new feature, and `yarn test` passes.
- [ ] I have detailed the new feature in the relevant documentation.
- [ ] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [x] If this is a breaking change I've explained why.

### CODE REVIEW

**Comment by benjie** [comment]
> The types that are in the repo are IMO wrong. 

Can you explain this? This module can be imported as either `require('...')` or `require('...').default` and it includes the `__esModule` special non-enumerable property, so it should be importable in either way. It works fine from TypeScript independent of your settings (I think?)... Why do you feel it would be better to make a less convenient import, or require `esModuleInterop`?

https://github.com/graphile/pg-simplify-inflector/blob/1dd49a877b53901b6c5b15091fddcf9164a1ef73/index.js#L490-L493

**Comment by keithlayne** [comment]
I think I said "wrong" where I should probably have said nothing. I guess it's just my preferences speaking - I don't like default exports, so I think for CJS modules I'd rather have a named export that will work with both CJS and ESM. If you could `export =` and `export default` simultaneously it wouldn't be an issue.

I can't logically defend that statement. The types work, and your choice is pragmatic. Sorry for the noise.

**Comment by benjie** [comment]
I'm happy also adding `export PgSimplifyInflectorPlugin`; that's what we do with PostGraphile - it's both the `default` export, and also a named export.

File: index.js
Comment (benjie) [inline index.js]: Though TypeScript has added this at the top of the file; it's then overwritten by the `module.exports =` above. So we should add it back, or find a better way.

**Comment by benjie** [comment]
Ah; I think I see what you're saying now; because of how I made it really compatible in JS, TS can't really write it without either a breaking change or some hacks. Disappointing. I'm okay with putting some TypeScript hacks in; but as you say maybe using `// @ts-check` and some type hints is sufficient?

**Comment by keithlayne** [comment]
I've never really don't the `@ts-check` thing, but even just adding jsdoc comments with types will improve DX I think, at least in vscode.  I know you can add a `jsconfig.json` as well. I was thinking just the comments could go a long way towards making it easier to work on for the uninitiated, which was my goal. After all, this is probably a pretty low-touch project most of the time, and I don't know if it makes sense to add a build process.

File: index.js
Comment (keithlayne) [reply index.js]: Yeah I wasn't thinking and missed the implication of moving that up.

**Comment by keithlayne** [comment]
I think the easiest way to use typescript and preserve behavoir is probably to:

1. Turn off declaration emit and keep the hand-written version.
2. In the source file, use `export =` 
3. Tack on the `exports.default` and `Object.defineProperty` at the end.

I think this will keep everything else as-is while adding typechecking.  It's probably possible to do it with `export default` but  that seems more complicated. (Not sure, didn't really try)

**Comment by benjie** [comment]
That sounds reasonable. Let's do that.

**Comment by keithlayne** [comment]
Nope, was wrong.  Seems the best way is to not use TS's export mechanism at all, but to preserve the last few lines of the JS.  The reason is that it pushes whatever export line there is to the bottom, which overwrites everything else.  Interestingly, It still puts the `Object.define` at the top, even with no `export` statements, I guess it detects CJS patterns. Changes incoming.

**Comment by benjie** [comment]
From a quick scan I think this is an okay approach. It's going to be quite a while before I review it fully and merge it as I'm working on the PostGraphile HTTP server right now, and next up is PostGraphile's websockets server, and these are both big tasks. So let it sit with me for a few weeks; if you don't get a review by mid November definitely ping me :+1: 

(And if you haven't already, please make sure that I can push changes to this branch.)

**Comment by benjie** [comment]
Thanks for taking this on :raised_hands:

File: index.ts
Comment (benjie) [inline index.ts:21]: Assertion - we know `index` will exist because the regexp doesn't use `/g`

File: index.ts
Comment (benjie) [inline index.ts:23]: Assertions: every string matches this regexp, so suffixMatches is guaranteed; and index will exist because we're not using `/g`

**Comment by benjie** [comment]
I've added a couple of assertions to remove the TypeScript specific unnecessary assertions, removed the `assert` function as we don't need it, and refactored the code to output code identical to the previous code. I've also changed the TSConfig to target ES2018 so we can use continue to use rest-spread (previously index.js was full of `Object.assign` calls, but master uses object rest-spread), and have turned off source maps since we don't distribute index.ts in the package so there's little point in including it; plus index.js is so similar to index.ts a source map shouldn't be necessary. Finally I've ran prettier on the compiled TypeScript output, so we can see in the final diff that basically no code has changed whatsoever. :raised_hands:

**Review by benjie — APPROVED** [review]
`index.js` is basically unchanged, so this is an obvious merge - thanks for your work @keithlayne :raised_hands:

**Comment by keithlayne** [comment]
Thanks! I honestly didn't know the guarantees about the regex stuff.  I tend to err on the side of caution when the types are like that - I should have researched. Agree on all your points.  I think I picked ES2017 based possibly on the minimum node version I thought postgraphile supported, but I always have to look that up anyway :).

**Comment by benjie** [comment]
Yeah; technically PostGraphile supports Node 8; but this package doesn't :smiling_imp:

**Pull request merged**
